compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -790,19 +808,42 @@ class TypeComparer(initctx: Context) extends DotClass with ConstraintHandling {
   /** Subtype test for corresponding arguments in `args1`, `args2` according to
    *  variances in type parameters `tparams`.
    */
-  def isSubArgs(args1: List[Type], args2: List[Type], tparams: List[ParamInfo]): Boolean =
+  def isSubArgs(args1: List[Type], args2: List[Type], tp1: Type, tparams: List[ParamInfo]): Boolean =
     if (args1.isEmpty) args2.isEmpty
     else args2.nonEmpty && {
-      val v = tparams.head.paramVariance
-      def isSub(tp1: Type, tp2: Type) = tp2 match {
-        case tp2: TypeBounds =>
-          tp2.contains(tp1)
+      val tparam = tparams.head
+      val v = tparam.paramVariance
+
+      def compareCaptured(arg1: Type, arg2: Type): Boolean = arg1 match {
+        case arg1: TypeBounds =>
+          val captured = TypeArgRef.fromParam(tp1, tparam.asInstanceOf[TypeSymbol])
+          isSubArg(captured, arg2)
+        case _ =>
+          false
+      }
+
+      def isSubArg(arg1: Type, arg2: Type): Boolean = arg2 match {
+        case arg2: TypeBounds =>
+          arg2.contains(arg1) || compareCaptured(arg1, arg2)
         case _ =>
-          (v > 0 || isSubType(tp2, tp1)) &&
-          (v < 0 || isSubType(tp1, tp2))
+          arg1 match {
+            case arg1: TypeBounds =>
+              compareCaptured(arg1, arg2)
+            case _ =>
+              (v > 0 || isSubType(arg2, arg1)) &&
+              (v < 0 || isSubType(arg1, arg2))
+          }
+      }
+
+      val arg1 = args1.head
+      val arg2 = args2.head
+      isSubArg(arg1, arg2) || {
+        // last effort: try to adapt variances of higher-kinded types if this is sound.
+        // TODO: Move this to eta-expansion?
+        val adapted2 = arg2.adaptHkVariances(tparam.paramInfo)
```

Comment:
Complication: Need to adapt higher-kinded variances when comparing applied types as well as refined types.